### DISCUSSION TITLE
Potential issue on the multithreaded LA implementation

### DISCUSSION THREAD

**Discussion by hrkalona** [5E-324/Imagina#6] [General]
this part of the code:

``` c++
if (ThreadID == ThreadCount - 1) {
	LAI_.StepLength = j - PeriodBegin;
	LAs[ThreadID].Append(LA_);
	LAIs[ThreadID].Append(LAI_);
}
```

should be changed so the last thread to add those data should be the one who has `PeriodBegin != 0` (this variable should be initialized to 0) and the next thread (if any) must have have `PeriodBegin == 0`. If the last thread has `PeriodBegin !=0` then this thread should add the data.

This needs to be added so that the actual last thread that found the period should be the one to finalize the data.
This check should be added on the starter thread (thread 0) as well, so that the multithreaded code works even if you set the ThreadCount to 1.

Probably for this fix you need to add a new array, similar to Start, which stores the PeriodBegin for each thread, and to be initialized at -1, so you can do busy wait while the value is < 0.

**Comment by 5E-324** [comment]
Thanks for the suggestion. This function is experimental and is known to have many potential issues. However, since it has relatively low performance benefit, it's currently unused and its development is discontinued, so there's no plan to fix these issues. But if you have an implementation of the fix, feel free to submit a pull request.

**Comment by hrkalona** [comment]
In high period areas the LA calculation tends to be very slow, especially the first stage. That is why I was experimenting with your multithreaded code. I also sent you another potential issue/fix on the forums. If you think that both suggestions are ok I can raise a pull request.

If you also know any other unsolved issues let me know so I can debug them.

**Comment by 5E-324** [comment]
I replied to the message you sent me on the forums, and proposed a solution.

The problem with this function is mainly the handling of edge/corner cases. Here are some situations I can think of that are potentially problematic. Some of them may be duplicated or can be solved by the same fix. Further investigations are needed.

- A thread failed to find a dip in the part it's responsible for.
	- ... And searched to the end of the reference?
	- ... And didn't create any LA steps?
	- ... And found a dip elsewhere?
	- ... And created a LA step at the wrong place, maybe duplicate with a step created by the next thread?
	- ... Then what will happen to the previous thread?
- The last thread failed to find a dip. (this issue)
- Two threads doesn't agree on whether or not an iteration counts as a dip (related to the issue you mentioned on the forums)
	- The result of dip detection is influenced by previous iterations skipped by a step, but different threads starts at different iterations.
	- If so, the steps created by two threads will not line up.
	- The fix I proposed on the forums solves this, but in some rare occasions it's possible that none of the steps one thread creates lines up with any steps the next thread creates.

Some of these are more likely to occur when the orbit lacks of dips, which can happen if you zoom deeply into a Misiurewicz point, especially with the new version of period insertion, which I copied from the single threaded version in the last commit. (The old one will give up directly if no period is detected, which will make pixel computation slower) If you want to debug, consider including this kind of locations to your test cases.

Of course, not all of these are common or will cause serious issues, but without handling them, it will be less reliable than the single-threded version.

Is the single-threaded version slower than the high precision orbit computation? If not, then maybe it's better to let it run in parallel with the orbit computation. The multi-threaded version is intended for accelerating opening saved compressed references, if you don't need to do this, then maybe this function is also not needed.

Note: "dip" refers to the sudden drop of |Z|, it's what the function `DetectPeriod()` detects.

**Comment by hrkalona** [comment]
I will continue the discussion that I started on the forum here:

I had a scenario where a thread starting at j = 307 managed to detect a period at 1319,
so I got an exception: Thread( 5) error 1319 >= 375. 
The next thread (which started at 369) managed to find a period at 375 so this is why we got the exception.

I added a breakpoint on the single threaded code and I captured some indexes between them that the single threaded code managed to detect a period:

306, 308, 317, 319, 322, 324, 329, 334, 336, 345, 347, 350, 352, 357, 362, 364, 373, 375

so theoretically the thread starting at 307 could find any of these but it failed.
This made me think what would happen if instead of initializing a thread by doing a Step and incrementing j by +2, to not do a Step and increment by +1. I tested one work around:

``` c++
LAInfo<real> LA_ = LAInfo<real>(reference.Ref[j]);

if (LA_.DetectPeriod(reference.Ref[j + 1]) || j + 1 >= reference.RefIt) {
   j++;
}
else {
   LA_ = LA_.Step(reference.Ref[j + 1]);
   j += 2;
}
```

which seemed to work but you said that DetectPeriod is not designed to work like that and it might create some other edge cases.
Before I start working the other method you proposed, which requires a synchronization redesign what If I do this,

``` c++
LAInfo<real> LA_ = LAInfo<real>(reference.Ref[j]);
LAInfo<real> LA_2 = LAInfo<real>(reference.Ref[j]);
LA_2 = LA_2.Step(reference.Ref[j + 1]);

size_t j1 = j + 1;
size_t j2 + j + 2;
```

and then do the loop and and check both of them:

``` c++
LAInfo<real> NewLA;
bool PeriodDetected1 = LA_.Step(NewLA, reference.Ref[j1]);

LAInfo<real> NewLA2;
bool PeriodDetected2 = LA_2.Step(NewLA2, reference.Ref[j2]);
```

and basically break the loop when the first detection occurs, so the starting J will be either the value of j1 or j2

Does this makes any sense?

**Comment by hrkalona** [comment]
Actually I tested this and in some other cases it detected the period right away leading to other collisions.
I am gonna test something else. 
I am still going to have two different indexes j1 and j2 and two different starting LA_s (LA_ and LA2_) but I will run two independent starting search loops and let both of them try to detect a period.
When they are done I need to check if the original detection method which start with +2 offset managed to detect a period within its search space:
size_t Start = reference.RefIt * ThreadID / ThreadCount;
size_t End = reference.RefIt * (ThreadID + 1) / ThreadCount;

if the original method found something and it is between start and end then we will use that, else we need to check the secondary detection that started with an offset of +1. If this detection is between Start and End then we might be ok.

In any case If a thread detects a period outside its search space maybe we need  to stop this thread of doing anything else, and maybe let the previous one work its way out to cover the next thread's space as well.
Given that the starter thread is implement in a way that it can work one a single threaded execution then if all threads fail then the starter thread can work alone.

**Comment by hrkalona** [comment]
Or we can skip the idea with two indexes and two LA_s and just add this check after the first loop of seach:

``` c++
if (ThreadID == ThreadCount - 1) {
	Start[ThreadID] = j;
}
else if (j >= Begin && j < End) {
	Start[ThreadID] = j;
}
else {
	while (!Start[ThreadID + 1]);
	Start[ThreadID] = Start[ThreadID + 1];
	return;
}
```

If the last thread will not find a period then j will be >= reference.RefIt so we are covered. If any of the previous threads will not find anything, this value will be propagated up until thread 1, so thread 0 will do all the work.
If any of the intermediate threads fails to find a period, it will just pass the next threads value to its predecessor and then stop.
This value will either be an intermediate value or a value >= reference.RefIt, so its predecessor will do all the work up until that value.

**Comment by 5E-324** [comment]
Do you mean if the initial detection failed to find a dip in the range, then just let the previous thread to do its work? This should work in this case, but can lead to bad performance. It's probably better to try the following two methods.

If you want to do two initial detections, then maybe you can let one of them start one iteration before or after the other, like this:

```c++
LAInfo LA_ = LAInfo(reference.Ref[j - 1]);
LAInfo LA_2 = LAInfo(reference.Ref[j]); 
LA_ = LA_.Step(reference.Ref[j]);
LA_2 = LA_2.Step(reference.Ref[j + 1]);

size_t j1 = j + 1;
size_t j2 + j + 2;
```

Another method is to skip the initial detection completely, start the second loop right away, but discard the first few steps it creates. See commit aac0d3c8. It should be more robust than the original one. You can test it to see if it's really working as expected.
The more steps it discards, the more likely the start will line up with a step created by the previous thread. In most cases, discarding 1 to 3 steps should be enough. I temporarily hard-coded the number to be 2.

Note that none of the methods mentioned here is perfect. They won't completely solve the third issue I mentioned previously: "Two threads don't agree on whether or not an iteration counts as a dip", though this will not occur frequently.

By the way, you can put your code in code blocks, like this:
````
```
// your code here
```
````
Optionally, specify the language.
````
```c++
// c++ code
```
````

**Comment by hrkalona** [comment]
Also another good idea is to add some threshold in the beginning in order to use the single threaded variant when  reference.RefIt / ThreadCount is less that some factor.

**Comment by 5E-324** [comment]
Or use fewer threads. The number of threads should be the smallest of `reference.RefIt / MinIterationsPerThread`, `reference.RefIt / (Period * MinPeriodsPerThread)` and hardware concurrency, and is at least 1.

**Comment by hrkalona** [comment]
As for finalizing the data something like that can be used:

for Worker 0, after the loop add:
```c++
               
		finished[0] = true;

		int lastThreadId = findLastThreadId(finished, LAs, ThreadCount, 0);

		if (lastThreadId == -1) {
			return;
		}

		if (0 == lastThreadId) {
			LAI.StepLength = i - PeriodBegin;

			reference.LA.Append(LA);
			reference.LAI.Append(LAI);
		}

```

for worker i, after the loop add:
```c++

		finished[ThreadID] = true;

		int lastThreadId = findLastThreadId(finished, LAs, ThreadCount, ThreadID);

		if (lastThreadId == -1) {
			return;
		}

		if (ThreadID == lastThreadId) {
			LAI_.StepLength = j - PeriodBegin;

			LAs[ThreadID].Append(LA_);
			LAIs[ThreadID].Append(LAI_);
		}

```

The code for findLastThreadId:
```c++

		template<typename real>
		int findLastThreadId(bool finished[], ReservedMemoryArray<LAInfo<real>> LAs[], int ThreadCount, int ThreadId) {

			for (int i = ThreadCount - 1; i >= 0; i--) {
				while (!finished[i]);

				if (LAs[i].size() > 0) {
					return i;
				}
			}

			if (ThreadId == 0) {
				return 0;
			}

			return -1;

		}

```

The array of finished should be initialized to false during the array initialization for the thread data

**Comment by hrkalona** [comment]
By the way I have tested the new implementation on my code, using a large number of threads (128) as this was initially creating one of my edge cases.
On one location that the previous code was working now I get:
Thread( 1 ) error 8890 > 8888
so at some point PeriodBegin > Start[nextThread]
Maybe I need to increase the value of  DiscardCount?

Edit I changed it to 5 but it did not help, it just offseted the two values:
Thread( 1 ) error 8899 > 8897

**Reply by 5E-324** [reply]
Have you tried the other method I mentioned? Does it work?

**Reply by 5E-324** [reply]
Can you provide these information?
- The period.
- The boundary between the two threads.
- The dips these two threads have detected near these iterations, preferably also the result of the single-threaded version.
- |Z| near these iterations, preferably rounded to 2-3 significant figures.

**Reply by hrkalona** [reply]
I will rewrite the code with the DiscardCount and I will rerun it so I can provide the data.

**Reply by hrkalona** [reply]
[data.txt](https://github.com/5E-324/Imagina/files/14040118/data.txt)

**Reply by 5E-324** [reply]
So the error is different from the last time?
And |Z| never gets smaller than 0.019 in this range? (And this is 3 decimal places, not significant figures. It's fine for this location, though, but if |Z| ever gets smaller than 0.001, it will underflow. It's also fine to not round it, just use the default format. It will be harder to read, but there's no risk of losing information.)
Are you using the same location as last time? If so, then your previous code has a bug. I'm not sure about this one, though, but some of the data looks a bit weird...
Iteration 11 is clearly not a dip, so it must be an inserted period (which indicates Period > 64 or not detected). What's the original period, i.e. the period found by the first loop, before `if (Period == 0) {...}`?
Does the other method work at this location?

**Reply by hrkalona** [reply]
The error is from the same location, its just that I changed the detection thresholds to match the ones from Imagina, as I had changed them for my version. I wanted to be as accurate to the original Imagina code now, so we still got the exception but at different indexes.
The detected period from the first loop is 16880, so its > 64 and after this part of the code:
```c++
double NthRoot = round(log2(double(Period)) / 4); // log16
Period = round(pow(double(Period), 1.0 / NthRoot));
```
we get 11

The |Z| at 16880 is: Iteration: 16880 |Z|: 1.6005864523587292E-42

Here are the data without rounding:
[data.txt](https://github.com/5E-324/Imagina/files/14040626/data.txt)

The 2 detectors method (LA_ and LA_2) worked on this location

The location is at:
-1.7685653943536636812525937129345323689264178203655023808403568327813984751602057983694075544809385380635853233562272021595486038792597346267621026418533261924962244609714446430169626331467057579947033159779441985348128008304981334471441266997824980553144563791567792114724296121872944832614093450944724492334683449190475005836849397812504637683710510408861059046995537629068401998856337076707729082352934525561410799942777533219989198396355688230764560224323159046285245196155466314399804239661358274503819615557368332052446821888562141313581523317892262380088027074432482039394916527640029121955174213203521566050696042500969919946717868654263385845711621571084795576591138961066064076009303183875949501143195217469003173308513486183081233774459137115288547807735477262202007250717008710283162223251492449408111203225925774749231087606075791025786880373414419640567812051244300178691506999924356763498211814367907336500103294768040948403104863777932123346593677739181711573037377537612597105572567432844531248119698069826986109774175430124235000261952935148608089775593025955571455701166888054695292300393363716767974622037919908584052332982637466

0.00149693415390767795776818884840489556855946301445691471574014563855527433886417969977385819538260268120841953162872636930325763746322273045770475720864573841501787930094585669029854545526055550254240550601638349230447392478835897915689588386917873306732459133130195499040290663241163281171562214964938877814041525983714426684720617999806166857035264185620487882712073265176954914054913266203287997924901540871019242527521230712886590484380712839459054394699971951683593643432733875864612142164058384584027531954686991700717520592706134315477867770419967332102686480959769035927998828366145957010260008071330081671951130257876517738836139132327131150083875547829353693231330986024536074662266149266972020406424662729505261246207754916338512723205243386084554727716044392705072728590247105881028092304993724655676823686703579759639901910397135711042548453158584111749222905493046484296618244721966973379997931675069363108125568864266991641443350605262290076130999673222331940884558082142583551902556005768303536299446355536559649684565312212482597275388117026700207573378170627060834006934127513560312023382257072757055987599151386137785304306581858

Height: 6.0964e-1142
Iterations: 1500000

**Reply by 5E-324** [reply]
After investigating it a bit more, I think I found the reason. As I mentioned before, this method will break in some corner cases, this is one of such cases. This problem is caused by an imperfection of the HInf algorithm, and the coincidence that 243(an undetected true period) mod 11(the inserted period) = 1.
The first loop did not detect the period 243, which appears in the later parts of the orbit (and this period can't be detected, iteration 243 is not even a dip). It detects a longer period, and since the period is too long, it inserts a (fake) period 11 in an attempt to improve the performance. This can be problematic, as 243 mod 11 = 1, which means if a step successfully synchronized with the beginning of a period, then when it reaches the end of the period, the length of the newest step will be 1. Since the dip detection isn't meant to work when the step length is 1, it will create a step of length 2 without doing dip detection, so it may miss the end of the period. To avoid this issue, I added an extra check, which apparently doesn't work in this case. The |Z| at the dips (0.0187...) is small enough to be detected sometimes, but is not small enough to be detected consistently. Because the step used for the check is relatively long, it's relatively less sensitive, it also missed the end of the period, so in the next period, all steps start 1 iteration earlier relative to the beginning, and the last step in this period will have length 2. This step will be used for the dip detection, and since it's shorter, it's also more sensitive. It will detect the dip, and the process repeats. As a result, a dip will be detected every 2 periods (in your data, many adjacent dips are 486 iterations apart, which is 2 periods). This won't be a problem in single-threaded code, but in multi-threaded code, it's possible that one thread only detects the even periods and another thread only detects the odd periods, and they never align with each other.

The solutions/workarounds are:
- Adjust the thresholds and/or other parameters and see if this helps.
- Disable period insertion.
	Change

	```
	if (!Period) {
		if (reference.RefIt > 64) {
			...
		} else {
			...
		}
	} else if (Period > 64) {
		...
	}
	```

	To

	```
	bool InsertPeriod = false;
	if (!Period) {
		if (InsertPeriod && reference.RefIt > 64) {
			...
		} else {
			...
		}
	} else if (InsertPeriod && Period > 64) {
		...
	}
	```

	This may affect performance and is not guaranteed to work.
- Make it more permissive of misalignment.
- Use another method.

Since you've already got a working method, you can use it instead.

By the way, what are the thresholds you are using in your version?

**Reply by hrkalona** [reply]
I have the thresholds to be configurable, but I changed the default values to these:

- Stage0PeriodDetectionThreshold = 0.1;//Imagina: 0x1.0p-10;

- PeriodDetectionThreshold = 0.1;//Imagina: 0x1.0p-10;

- Stage0PeriodDetectionThreshold2 = 0.5; //Imagina: 0x1.0p-6

I have picked this values because I wanted to make the average Perturbation steps to almost match claude's version (at least on some location, with the expense of storing more steps).

Also I have experimented with this as well
```java
    private int getNthRoot(double val) {
        if(NthRootOption == 0) {//Original
            double NthRoot = Math.round(log2(val) / 4.0);
            NthRoot = NthRoot < 1 ? 1 : NthRoot;
            return  (int)Math.round(Math.pow(val, 1.0 / NthRoot));
        }

        double NthRoot = log2(val) / periodDivisor;
        NthRoot = NthRoot < 1 ? 1 : NthRoot;
        return  (int)Math.round(Math.pow(val, 1.0 / NthRoot));
    }
```

The second code allows for a configurable divisor and skips Math.round after the log

**Comment by hrkalona** [comment]
I have only tested the method with DiscardCount and the one that I described about making a thread quit if it does not find a start on its corresponding range.

I haven't tested the one with the 2 LAs that do the 2 initial steps with an offset of 1.
How would this last method look like? 1 Starter loop with 2 tests inside it? Whichever test works first should do? 

Or 2 starter loops? If the original loop finds a period between the threads range we use that else we run the second loop with the offseted step?

**Reply by 5E-324** [reply]
1 loop with 2 tests, same as in the second method you mentioned in https://github.com/5E-324/Imagina/discussions/6#discussioncomment-8221117

**Comment by hrkalona** [comment]
I have tested it and it seems to work, I am going to post a snippet of java just to show you what I did (just because it is easier do debug the code there)

```java
                        LAInfoI LAI_ = new LAInfoI();
                        LAI_.NextStageLAIndex = j;

                        GenericLAInfo LA_2 = GenericLAInfo.create(maxRefIteration, deepZoom, j, referenceDecompressor);
                        LA_2 = LA_2.Step(j + 1, referenceDecompressor);
                        int j2 = j + 2;

                        GenericLAInfo LA_ = GenericLAInfo.create(maxRefIteration, deepZoom, j - 1, referenceDecompressor);
                        LA_ = LA_.Step(j, referenceDecompressor);
                        int j1 = j + 1;


                        int PeriodBegin = 0; 
                        int PeriodEnd = 0; 
                        boolean PeriodDetected2 = false;
                        boolean PeriodDetected = false;

                        for (; j2 < maxRefIteration; j1++, j2++) {
                            GenericLAInfo NewLA2 = GenericLAInfo.create(maxRefIteration, deepZoom);
                            PeriodDetected2 = LA_2.Step(NewLA2, j2, referenceDecompressor);

                            if(PeriodDetected2) {
                                LAI_.NextStageLAIndex = j2;
                                PeriodBegin = j2;
                                PeriodEnd = PeriodBegin + Period;

                                int jp1 = j2 + 1;

                                if (jp1 >= maxRefIteration) {
                                    LA_2 = GenericLAInfo.create(maxRefIteration, deepZoom, j2, referenceDecompressor);
                                    j2++;
                                } else {
                                    LA_2 = GenericLAInfo.create(maxRefIteration, deepZoom, j2, referenceDecompressor).Step(jp1, referenceDecompressor);
                                    j2 += 2;
                                }
                                break;
                            }

                            GenericLAInfo NewLA = GenericLAInfo.create(maxRefIteration, deepZoom);
                            PeriodDetected = LA_.Step(NewLA, j1, referenceDecompressor);

                            if(PeriodDetected) {
                                LAI_.NextStageLAIndex = j1;
                                PeriodBegin = j1;
                                PeriodEnd = PeriodBegin + Period;

                                int jp1 = j1 + 1;

                                if (jp1 >= maxRefIteration) {
                                    LA_ = GenericLAInfo.create(maxRefIteration, deepZoom, j1, referenceDecompressor);
                                    j1++;
                                } else {
                                    LA_ = GenericLAInfo.create(maxRefIteration, deepZoom, j1, referenceDecompressor).Step(jp1, referenceDecompressor);
                                    j1 += 2;
                                }
                                break;
                            }
                            LA_2 = NewLA2;
                            LA_ = NewLA;
                        }

                        if(PeriodDetected2) {
                            LA_ = LA_2;
                            j = j2;
                        }
                        else if(PeriodDetected) {
                            j = j1;
                        }
                        else {
                            j = maxRefIteration;
                        }

                        Start[ThreadID] = j;
```

The only issue is that since j2 is always one step ahead it will stop the loop while j1 can still do one check (If we have not found the period yet)

**Comment by 5E-324** [comment]
> The only issue is that since j2 is always one step ahead it will stop the loop while j1 can still do one check (If we have not found the period yet)

Maybe it's better to let LA_ do a test first, then they will be synchronized, and you won't need j2 anymore, and the two if statements inside the loop can be merged.

Edit: Or simply reorder the two tests in the loop.

**Reply by hrkalona** [reply]
If I do the LA_ check first (which essentially starts at j + 1) we get two possibilities:

- it either finds the period immediately and breaks
- it does not find the period and at that point LA_ synchronized with LA_2 so we do not need LA_ anymore

Does this mean that we can just use the initial code without having 2 checks and just do this?
```c++
LAInfo LA_ = LAInfo(reference.Ref[j - 1]);
LA_ = LA_.Step(reference.Ref[j]);
j++;
```

basically start with a step one index earlier?

**Reply by hrkalona** [reply]
Scratch that, I re-arranged the checks so the step and check for LA_ happens first. I checked the NewLA value and compared with the LA_2 before executing the second step, and its internal values were not the same, so they did not synchronize. So the idea that I proposed above does not work.

**Reply by 5E-324** [reply]
> - it either finds the period immediately and breaks

Does this also happen to LA_2, or it only happens to LA_1, or is this just a hypothesis?

> - it does not find the period and at that point LA_ synchronized with LA_2 so we do not need LA_ anymore

They start at different iterations, which can make a difference.

> Does this mean that we can just use the initial code without having 2 checks and just do this?
> 
> ```c++
> LAInfo LA_ = LAInfo(reference.Ref[j - 1]);
> LA_ = LA_.Step(reference.Ref[j]);
> j++;
> ```
>
> basically start with a step one index earlier?

Isn't it the same as shifting the thread boundary by 1? It just let a different set of cases to be the edge cases.

BTW, would you like me to diagnose the error you encountered when using the other method? If so, I will need the information shown in https://github.com/5E-324/Imagina/discussions/6#discussioncomment-8228926

**Comment by hrkalona** [comment]
Ok I will give it a try.
By the way the actual reason that I am pushing this investigation is mainly because of using a compressed reference.
The reference needs to be decompressed on the fly so it increases the work that needs to be done during the LA calculation.
Maybe this is the only reason that a multi-threaded variant makes sense.

**Comment by hrkalona** [comment]
By the way, this is not relevant to the ticket, but on my testing location:

-0.7835725813169320184840224867376351092924750527283126605770458611370455267265454388204427085809245638198799454654985287669285990956175539126535519949740449204 122886107680174962376414339939088848820857226070485123924786941622746796123165944402589177219956098143276844795046628269028399104283451871817155302216960926282 363499594314655253587032982335629132577003672867861106333407748103438849706873099526939225857669635066780920077649297961547281611992915833569164229737882549238 092687788048705100724593747882302897187445650536749535118020303540502881021797054418537443630317884153146247701763485079573562673125418195039451666236192703721 907412635653222752265439837424603127681448446902281221234987718499182189953727481372418823050005537632211394694354859499307127329547326194540666288999857484529 479990216227582475525866441056476024471638468476338381292546042145336696347162859110653321944644205811495484952048087995050138388459532060820252555240674533882 893672253614217241924876759554079560775344043892599868208711295674413699576061525437044124119749608673246785739229651682336366705530400807638670656054599653496 760851763951358419276469686859251026924456274112250471828721733721408332313124625509258852885816469866133806686514370775790460537363963729181750528503160362342 432598440656978661725829621526922746473529283910928888846871988485123730121355189309838105681199817897378173789168393428726775635828928596636398887965993049289 707264868684637847581831311833527276451683985821938222293793061913525665213301230007852438653273233030119052783095641871330002918801136639799069129138907229426 256048873221492887968000777447039775617704093782625858808906130060898057733228917723600083002033142983569500169521744174221983551804089092618467707285531337347 601204225399420398495590807328213279207735738313415447521316439368934843913032405410037606721215498180115894449295102254548062289628439937834345306800710766239 856961056459444706354512855497853008951101134176995184347381818937159441059431247776808934927545741985188140501073097558253437462645154279435424954534710147178 747646663860403869595665804009150958182118526567283618876818676132665341218560426980184384040725267381796822357879807670462344848783056800194985704734642439582 573474552660576181882618746443766307568570878854958885504313083820191618948927559733736400763266871662219508663391344883528106673297458584254804916918107200012 484034055319239194261921784352885725010350132095248752245384465782231282738820162372326774215797140581737801100656833116739880548116374119323299472810437557314 037252128532114920048331893339041090270397821017770969433557859110485116268851144746296033230491224227267852437136157501882605027031190975699867347374181390670 290229475146021329060794388149139368171393189852757651729518247574797190306006528002726153505863253484140598717715618513549226634151269275213483385372558531879 691939622512372684211879768218254963764458461755706179886504812228259718437175904605561931876306552924848779636953193219678524340015001122350464156554551268169 8453040738076250439068398583960345899567892815087977308822415484944086025031863

0.1409714759189446566817605846018091238514576578254413743698490019628233600132106201216722291648146498319305545842609072928997223027252192901095510553698197321 717637262074213915811017477639605444393870648505148991099948456085734212337610048882701120369652883270791476993112596980842524188605039153487905917261879546226 583794237331485400580356523698574249644328449040660242344535651344783593190381096670179475625618993933483912997227226243192549273177135313282365766660977140145 614829713868939260982459426046138533812733179154957888387644708469347800633199261311106237055380682505252276384872810372701022255159246437534681828490623546373 226264389727041409793546568063790456416211750802618079839288308189713667743894489728860218509087439190332383807598175807344252153567867241713809331811304947632 753425395237164239836724896467084867599284893894378672826506678282641456266864054002996570678013955536040221987348615063035436838144153864801035134601167523292 007083625625453981424570438965010604104540353919934158353539701263809334259160068804206620722512044654532150840916632951598242131124324289995145331977950207183 789756737551361890771234370704800760460593807662619091029212481152844551171840909505622772532033348552459248727192756691002679296480087830825929365435642140780 684805008536062000957204322869985227661129244146186460854291124482604967876137765447490806439650445269342817105788731638623495411626650523012543645647830758731 266150330842856241753523408185892128591296241020208950593674089349116244915166725912494821533619540904921251789496210538436243729177707234801982510438188021748 054170192423026130084545418803021599414048964974265360918273111657803510552653749532672008479280926059281935491998919397292951524324307935968667095132607759091 196065433004296148521664532583447678774877921407755029849168659312848355672244209244796518218618917841320056225128378992705494496587521674754795891205548672426 644579609051140177839985041245140665298518515825489205523397817857314549636520101182068202237111468179558671339029984942912204542829593179165375830059065533709 008441916095067325641226023300985985645179142465691732311822910766225843696550562784747366100267073133483163262181538370571148344572923394627039552481617048141 568915522062486262650844872315960576491128976174670754270909325017438378931200175555734522415037667582387651105974325062818636594633556749254780797884743239179 119915758702008536400413977596825609344256923249706643920475544029341118927328857173726900518428260192329164958157269003798402242435177914107999978162255029059 432276699338005022703804998315272112898204661894836658135021505657091094056085006097447617431008557662962945322420378607456028127594225293313119896651920661678 313538148918861661458528208935935479360453622995002935004983781081606681271529806054837277758966445480809766283899977416151716113620023975824375068076697286749 544083271016251127598772639762508517927841957403649248287596598120028505928758821713672020445684993888707575811227949570771568819802222834253928564199425639029 3626704096006142686191615927829997121552009564979821977515676624039388572675029

height: 1.9e-3068

iterations: 536870912

Imagina takes about 47min to calculate the reference. My java wrapped mpir implementation is multithreaded and it takes about 17mins.
So there are places that having a multithreaded reference implementation  is of great benefit.
Here is the 3 squares variant using 3 threads in case you want to try out at some point:
https://github.com/hrkalona/Fractal-Zoomer/blob/release/1.0.9.0/src/fractalzoomer/core/mpir/custom_mpir.c#L75

**Comment by hrkalona** [comment]
I have found a different edge case now but lets sum up what I have done.

- The first synchronization loop uses 2 detectors start at j-1,j and j,j+1
- After the loop I do this:
```java
                        if(ThreadID == LastThread || (j >= Begin && j < End)) {
                            StartIndex[ThreadID].complete(j);
                        }
                        else {
                            int nextStart = StartIndex[nextThread].get();

                            if(nextStart < 0) {
                                return;
                            }

                            StartIndex[ThreadID].complete(nextStart); //Abort the current thread and leave its task to the previous
                            return;
                        }
```
I make sure that each thread begins on its range. It usually does not happen, but I saw it happen at 0+1i where all threads failed and the starter thread did all the work.

- The last thread to finalize the data is the one which added at least one entry on this thread local data

Even with those fixes I have got your final edge case.
Thread( 14 ) error 1128 >= 1125
So I guess two consecutive threads are not aligned anymore.

Quoting what you told me originally:


**_A better solution is to let the previous thread continue creating new steps after passing the end, until a step ends exactly at the beginning of a step created by the next thread._** 

Because it is challenging for me to read the other threads local data in order to figure out the alignment, and stop the calculation on an arbitrary position I am trying something else:
```java
                               if (j > End) {
                                    if (ThreadID == LastThread) throw new Exception("Last Thread cannot reach this");

                                    if(j >= maxRefIteration) {
                                        break;
                                    }

                                    if(nextThread < StartIndex.length) {
                                        int nextStart = StartIndex[nextThread].get();

                                        if (nextStart < 0) {
                                            return;
                                        }

                                        if (j == nextStart - 1) {
                                            j++;
                                            break;
                                        } else if (j >= nextStart) {
                                            //throw new Exception("Thread( " +ThreadID + " ) error " + j + " >= " + nextStart);
                                            System.out.println("Thread( " + ThreadID + " ) error " + j + " >= " + nextStart);
                                            nextThread++;
                                        }
                                    }
                                }
```

Basically when an exception is meant to be thrown I continue and from now on I start checking the next-next threads start index just in case we align with that. This means that during the concatenation I will have to dismiss one thread's (or more data).
This code needs to be added to the Thread 0 as well. The worst case scenario is that the thread 0 does not find any alignment at all with the next threads, so it has to calculate everything. The other threads will do their calculations as well but they will be thrown away.

With this solution I also have to modify the final data inclusion code. I now do it differently.
Each threads calculates the last value and instead of adding it to the local thread data, it stores it into a different container.
```java
                        LAI_.StepLength = j - PeriodBegin;

                        LAData laData = new LAData();
                        laData.la = LA_;
                        laData.setInfo(LAI_);

                        LastLAPerThread[ThreadID] = laData;
```

during the concatenation I do the following:
```java
     {
            int lastThreadToAdd = 0;

            int i = 0;
            int j = i;
            while (i < Tasks.length - 1 && FinishIndex[j] > StartIndex[i + 1].get()) {
                i++; //Skip, if there is a missalignment
            }
            i++;

            for (; i < Tasks.length; i++) {
                int length = LAcurrentIndexPerThread[i];
                LAData[] threadData = LAsPerThread[i];
                for (int k = 0; k < length; k++) {
                    addToLA(threadData[k]);
                }

                if(FinishIndex[i] > StartIndex[i].get()) { //If this thread managed to search
                    lastThreadToAdd = i;
                }

                j = i;
                while (i < Tasks.length - 1 && FinishIndex[j] > StartIndex[i + 1].get()) {
                    i++; //Skip, if there is a missalignment
                }
            }

            addToLA(LastLAPerThread[lastThreadToAdd]);
        }
```

- Note that we now need to store the last index of each thread.
- Also the data finalization was changed because of another edge case. Its not longer the last thread that added any data, but the last thread that did a search in order to find the data (I found a case that the last thread added nothing, but it did search, so its EndIndex > StartIndex)

The StartIndex was modified to be a CompletableFuture in order to avoid busy wait with while loops.

Now back to the edge case that triggered this:
Thread( 14 ) error 1128 >= 1125

Thread 14 Start at 1050, Thread 14 Should have ended at 1125 but it failed.
Thread 15 Start at 1125,  End at 1202
Thread 16 Start at 1202,  End at 1276

After letting thread 14 keep going, it was now checking to align with thread 16 which starts at 1202, and in fact it did!
So during concatenation the code dismissed Thread's 15 data. 

On the same location, when I set the thread count to 128, it managed to work when two consecutive threads where missaligned:
Thread( 94 ) error 1123 >= 1120
Thread( 94 ) error 1128 >= 1125

So it now looks promising. We might be wasting some computations, but for cases where the first stage computation is long, it might be worth it.
